lib/export/dwca/data.rb

```diff
@@ -629,37 +887,664 @@ def biological_associations_resource_relationship_tmp
+        SQL
+
+        # Get images via observations
+        image_ids += conn.execute(<<-SQL).values.flatten
+          SELECT DISTINCT images.id
+          FROM (#{co_sql}) AS co
+          INNER JOIN observations obs ON obs.observation_object_id = co.id
+            AND obs.observation_object_type = 'CollectionObject'
+          INNER JOIN depictions ON depictions.depiction_object_id = obs.id
+            AND depictions.depiction_object_type = 'Observation'
+          INNER JOIN images ON images.id = depictions.image_id
+        SQL
+
+        # Get direct sounds (via conveyances)
+        sound_ids += conn.execute(<<-SQL).values.flatten
+          SELECT DISTINCT sounds.id
+          FROM (#{co_sql}) AS co
+          INNER JOIN conveyances ON conveyances.conveyance_object_id = co.id
+            AND conveyances.conveyance_object_type = 'CollectionObject'
+          INNER JOIN sounds ON sounds.id = conveyances.sound_id
+        SQL
+      end
+
+      if @media_extension[:field_occurrences]
+        fo_sql = @media_extension[:field_occurrences]
+
+        # Get direct images (via depictions)
+        image_ids += conn.execute(<<-SQL).values.flatten
+          SELECT DISTINCT images.id
+          FROM (#{fo_sql}) AS fo
+          INNER JOIN depictions ON depictions.depiction_object_id = fo.id
+            AND depictions.depiction_object_type = 'FieldOccurrence'
+          INNER JOIN images ON images.id = depictions.image_id
+        SQL
+
+        # Get images via observations
+        image_ids += conn.execute(<<-SQL).values.flatten
+          SELECT DISTINCT images.id
+          FROM (#{fo_sql}) AS fo
+          INNER JOIN observations obs ON obs.observation_object_id = fo.id
+            AND obs.observation_object_type = 'FieldOccurrence'
+          INNER JOIN depictions ON depictions.depiction_object_id = obs.id
+            AND depictions.depiction_object_type = 'Observation'
+          INNER JOIN images ON images.id = depictions.image_id
+        SQL
+
+        # Get direct sounds (via conveyances)
+        sound_ids += conn.execute(<<-SQL).values.flatten
+          SELECT DISTINCT sounds.id
+          FROM (#{fo_sql}) AS fo
+          INNER JOIN conveyances ON conveyances.conveyance_object_id = fo.id
+            AND conveyances.conveyance_object_type = 'FieldOccurrence'
+          INNER JOIN sounds ON sounds.id = conveyances.sound_id
+        SQL
+      end
+
+      image_ids.uniq!
+      sound_ids.uniq!
+
+      Rails.logger.debug "dwca_export: found #{image_ids.count} images and #{sound_ids.count} sounds to export"
+
+      { image_ids: image_ids, sound_ids: sound_ids }
+    end
+
+    # Create temporary tables with pre-computed API links for images and sounds
+    # This must use Ruby because API link generation uses the URL shortener
+    # @param image_ids [Array<Integer>]
+    # @param sound_ids [Array<Integer>]
+    def create_media_api_link_tables(image_ids, sound_ids)
```

Comment:
This is a very long method.  Feels like we could split some parts out to individual functions.